controllers/clustergroupupgrade_controller.go

```diff
@@ -315,86 +313,83 @@ func (r *ClusterGroupUpgradeReconciler) Reconcile(ctx context.Context, req ctrl.
+
+							currentBatchTimeout := utils.CalculateBatchTimeout(
+								clusterGroupUpgrade.Spec.RemediationStrategy.Timeout,
+								len(clusterGroupUpgrade.Status.RemediationPlan),
+								clusterGroupUpgrade.Status.Status.CurrentBatch,
+								clusterGroupUpgrade.Status.Status.CurrentBatchStartedAt.Time,
+								clusterGroupUpgrade.Status.Status.StartedAt.Time)
+
+							if time.Since(clusterGroupUpgrade.Status.Status.CurrentBatchStartedAt.Time) > currentBatchTimeout {
+								// Check if this was a canary or not
+								if len(clusterGroupUpgrade.Spec.RemediationStrategy.Canaries) != 0 &&
+									clusterGroupUpgrade.Status.Status.CurrentBatch <= len(clusterGroupUpgrade.Spec.RemediationStrategy.Canaries) {
+									r.Log.Info("Canaries batch timed out")
+									meta.SetStatusCondition(&clusterGroupUpgrade.Status.Conditions, metav1.Condition{
+										Type:    "Ready",
+										Status:  metav1.ConditionFalse,
+										Reason:  "UpgradeTimedOut",
+										Message: "The ClusterGroupUpgrade CR policies are taking too long to complete",
+									})
+								} else {
+									r.Log.Info("Batch upgrade timed out")
+									clusterGroupUpgrade.Status.Status.CurrentBatchStartedAt = metav1.Time{}
+									if clusterGroupUpgrade.Status.Status.CurrentBatch < len(clusterGroupUpgrade.Status.RemediationPlan) {
+										clusterGroupUpgrade.Status.Status.CurrentBatch++
+									}
 								}
 							}
 						}
 					}
-				}
-
-				if isUpgradeComplete {
-					meta.SetStatusCondition(&clusterGroupUpgrade.Status.Conditions, metav1.Condition{
-						Type:    "Ready",
-						Status:  metav1.ConditionTrue,
-						Reason:  "UpgradeCompleted",
-						Message: "The ClusterGroupUpgrade CR has all clusters compliant with all the managed policies",
-					})
-				} else if !clusterGroupUpgrade.Status.Status.StartedAt.IsZero() && time.Since(clusterGroupUpgrade.Status.Status.StartedAt.Time) > time.Duration(clusterGroupUpgrade.Spec.RemediationStrategy.Timeout)*time.Minute {
-					meta.SetStatusCondition(&clusterGroupUpgrade.Status.Conditions, metav1.Condition{
-						Type:    "Ready",
-						Status:  metav1.ConditionFalse,
-						Reason:  "UpgradeTimedOut",
-						Message: "The ClusterGroupUpgrade CR policies are taking too long to complete",
-					})
-
+				} else {
+					// On last batch, check all batches
+					if r.isUpgradeComplete(ctx, clusterGroupUpgrade) {
+						meta.SetStatusCondition(&clusterGroupUpgrade.Status.Conditions, metav1.Condition{
+							Type:    "Ready",
+							Status:  metav1.ConditionTrue,
+							Reason:  "UpgradeCompleted",
+							Message: "The ClusterGroupUpgrade CR has all clusters compliant with all the managed policies",
+						})
+						nextReconcile = requeueImmediately()
+					} else if !clusterGroupUpgrade.Status.Status.StartedAt.IsZero() && time.Since(clusterGroupUpgrade.Status.Status.StartedAt.Time) > time.Duration(clusterGroupUpgrade.Spec.RemediationStrategy.Timeout)*time.Minute {
```

Comment:
better to use time.Before() instead of doing the conversation. check this 

https://pkg.go.dev/k8s.io/apimachinery/pkg/apis/meta/v1#Time.Before

Reply:
Not sure how this function is better than Since()